spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/ParquetWithSparkSchemaVisitor.java

```diff
@@ -182,18 +186,18 @@ private static <T> T visitField(
 
   private static <T> List<T> visitFields(
       StructType struct, GroupType group, ParquetWithSparkSchemaVisitor<T> visitor) {
-    StructField[] sFields = struct.fields();
-    Preconditions.checkArgument(
-        sFields.length == group.getFieldCount(), "Structs do not match: %s and %s", struct, group);
+    Map<String, StructField> sFieldsByName =
```

Comment:
@rdblue I'm not crazy about this change, but I don't think there are other options. Previously we could assume and enforce that the Spark struct, and the Parquet group had the same number of fields. With the UnknownType, the Spark Dataframe will have a NullType, but not corresponding field in Parquet.

Since the Spark dataframe does not carry field-IDs, we have to map it by name.

Reply:
One potential alternative. Parquet `GroupType` already has `Map<String, Integer> indexByName`. We can still iterate the spark fields. if the type is null, don't query the parquet type. Otherwise, call the `GroupType#getFieldIndex(String)` to get the field index from parquet.

Reply:
That's a great suggestion, I've incorporated it in the code. I'll let @rdblue decide on what he prefers :)